File: ubel-node/src/ruby_runner.js
// RubyBundlerScanner.js
import fs   from "fs";
import path from "path";

/**
 * Scans Ruby / Bundler projects for installed gems.
 *
 * Detection strategy:
 *   1.  A directory is a Bundler project root when it contains a Gemfile
 *       AND a Gemfile.lock (the locked, installed state).
 *   2.  Installed gems are read exclusively from Gemfile.lock (the resolved graph).
 *       Gemfile is parsed only for group classification (dev / test / prod).
 *   3.  PURL: pkg:gem/<name>@<version>
 *
 * Gemfile.lock structure:
 *   GEM
 *     remote: https://rubygems.org/
 *     specs:
 *       activesupport (7.0.4)
 *         concurrent-ruby (~> 1.0, >= 1.0.2)
 *         ...
 *   PLATFORMS
 *     ...
 *   DEPENDENCIES
 *     rails (~> 7.0)
 *     rspec-rails (>= 0, development)
 *
 * Gemfile group syntax:
 *   gem 'rails'
 *   gem 'rspec', group: :test
 *   group :development, :test do
 *     gem 'byebug'
 *   end
 */
export class RubyBundlerScanner {

  static inventoryData = [];

  // ─────────────────────────────
  // PURL
  // ─────────────────────────────
  static _gemPurl(name, version) {
    return `pkg:gem/${name.toLowerCase()}@${version ?? ""}`;
  }

  // ─────────────────────────────
  // Detect Bundler project root
  // ─────────────────────────────
  static _isBundlerRoot(dir) {
    return (
      fs.existsSync(path.join(dir, "Gemfile")) &&
      fs.existsSync(path.join(dir, "Gemfile.lock"))
    );
  }

  // ─────────────────────────────
  // Parse Gemfile.lock
  // Returns Map<lowercaseName, { name, version, dependencies[] }>
  // ─────────────────────────────
  static _parseGemfileLock(lockPath) {
    let content;
    try {
      content = fs.readFileSync(lockPath, "utf8");
    } catch {
      return new Map();
    }

    const index = new Map();   // lowercase name → entry

    // Sections are separated by blank lines with uppercase headers.
    // We only care about the GEM / PATH / GIT specs sections.
    let inSpecs = false;

    // Track current gem being parsed (for dependency lines)
    let currentGem = null;

    for (const rawLine of content.split("\n")) {
      const line = rawLine;

      // Section headers
      if (/^(GEM|PATH|GIT)$/.test(line.trim())) {
        inSpecs = false;
        currentGem = null;
        continue;
      }
      if (line.trim() === "specs:") {
        inSpecs = true;
        currentGem = null;
        continue;
      }
      // Exit specs block on next section header or PLATFORMS/DEPENDENCIES/BUNDLED WITH
      if (/^[A-Z]/.test(line) && line.trim() !== "specs:") {
        inSpecs = false;
        currentGem = null;
        continue;
      }

      if (!inSpecs) continue;

      // Indentation tells us structure:
      //   4 spaces  = top-level gem:  "    activesupport (7.0.4)"
      //   6 spaces  = dependency of current gem: "      concurrent-ruby (~> 1.0)"
      const indent = line.match(/^(\s*)/)[1].length;
      const trimmed = line.trim();
      if (!trimmed) continue;

      if (indent === 4) {
        // Top-level gem line: "name (version)"
        const m = trimmed.match(/^([A-Za-z0-9_.-]+)\s+\(([^)]+)\)/);
        if (!m) { currentGem = null; continue; }

        const name    = m[1];
        const version = m[2].split(", ")[0];  // take first version if multiple
        const key     = name.toLowerCase();

        currentGem = { name, version, dependencies: [] };
        index.set(key, currentGem);

      } else if (indent >= 6 && currentGem) {
        // Dependency line: "dep-name (constraint)"  or just "dep-name"
        const m = trimmed.match(/^([A-Za-z0-9_.-]+)/);
        if (m) currentGem.dependencies.push(m[1].toLowerCase());
      }
    }

    return index;
  }

  // ─────────────────────────────
  // Parse Gemfile for group info
  // Returns { prod: Set<string>, dev: Set<string> }
  // (lowercase gem names)
  // ─────────────────────────────
  static _parseGemfileGroups(gemfilePath) {
    const prod = new Set();
    const dev  = new Set();

    let content;
    try {
      content = fs.readFileSync(gemfilePath, "utf8");
    } catch {
      return { prod, dev };
    }

    const DEV_GROUPS = new Set(["development", "test", "staging"]);

    let currentGroups = [];   // groups active for the current block

    for (const rawLine of content.split("\n")) {
      const line    = rawLine.trim();
      if (!line || line.startsWith("#")) continue;

      // group :development, :test do
      const groupBlock = line.match(/^group\s+(.*?)\s+do$/);
      if (groupBlock) {
        currentGroups = groupBlock[1]
          .split(",")
          .map(g => g.trim().replace(/^:/, "").toLowerCase());
        continue;
      }

      // end  – close group block
      if (line === "end") {
        currentGroups = [];
        continue;
      }

      // gem 'name', ...
      // gem "name", group: :test
      // gem "name", groups: [:development, :test]
      const gemLine = line.match(/^gem\s+['"]([^'"]+)['"](.*)/);
      if (!gemLine) continue;

      const gemName = gemLine[1].toLowerCase();
      const rest    = gemLine[2] ?? "";

      // Inline group: group: :test  OR  groups: [:development, :test]
      const inlineGroup = rest.match(/groups?:\s*(\[?[^,\]]+\]?)/);
      let groups = [...currentGroups];

      if (inlineGroup) {
        const gStr   = inlineGroup[1].replace(/[\[\]]/g, "");
        const extras = gStr.split(",").map(g => g.trim().replace(/^:/, "").toLowerCase());
        groups = [...new Set([...groups, ...extras])];
      }

      const isDev = groups.some(g => DEV_GROUPS.has(g));
      if (isDev) dev.add(gemName);
      else       prod.add(gemName);
    }

    return { prod, dev };
  }

  // ─────────────────────────────
  // Scan a single Bundler project
  // ─────────────────────────────
  static _scanProject(projectRoot) {
    const index = this._parseGemfileLock(path.join(projectRoot, "Gemfile.lock"));
    if (!index.size) return [];

    const components = [];

    for (const [key, { name, version, dependencies }] of index.entries()) {
      const id = this._gemPurl(name, version);

      const resolvedDeps = dependencies.map(dep => {
        const resolved = index.get(dep);
        return resolved
          ? this._gemPurl(resolved.name, resolved.version)
          : this._gemPurl(dep, "");
      });

      components.push({
        id,
        name:         key,
        version,
        type:         "library",
        license:      "unknown",
        ecosystem:    "ruby",
        state:        "undetermined",
        scopes:       [],
        dependencies: resolvedDeps,
        paths:        [projectRoot],
        project_root: projectRoot
      });
    }

    return components;
  }

  // ─────────────────────────────
  // Assign scopes (prod / dev)
  // via BFS from Gemfile group declarations
  // ─────────────────────────────
  static _assignScopes(inventory) {
    const byId    = new Map(inventory.map(c => [c.id, c]));
    const nameIdx = new Map();

    for (const comp of inventory) {
      if (!Array.isArray(comp.scopes)) comp.scopes = [];
      const comps = nameIdx.get(comp.name) ?? [];
      comps.push(comp);
      nameIdx.set(comp.name, comps);
    }

    const projectGroups = new Map();
    for (const comp of inventory) {
      const root = comp.project_root;
      if (!projectGroups.has(root)) projectGroups.set(root, []);
      projectGroups.get(root).push(comp);
    }

    for (const [projectRoot, comps] of projectGroups.entries()) {
      const { prod, dev } = this._parseGemfileGroups(
        path.join(projectRoot, "Gemfile")
      );

      const propagate = (names, scope) => {
        const queue = [];
        for (const n of names) {
          for (const c of (nameIdx.get(n) ?? [])) {
            if (c.project_root === projectRoot) queue.push(c);
          }
        }
        const visited = new Set();
        while (queue.length) {
          const c = queue.shift();
          if (visited.has(c.id)) continue;
          visited.add(c.id);
          if (!c.scopes.includes(scope)) c.scopes.push(scope);
          for (const dep of c.dependencies) {
            const d = byId.get(dep);
            if (d && d.project_root === projectRoot) queue.push(d);
          }
        }
      };

      propagate(prod, "prod");
      propagate(dev,  "dev");

      // Fallback – unscoped gems default to prod
      for (const c of comps) {
        if (c.scopes.length === 0) c.scopes.push("prod");
      }
    }
  }

  // ─────────────────────────────
  // Merge duplicates by PURL
  // ─────────────────────────────
  static mergeInventoryByPurl(components) {
    const map = new Map();

    for (const comp of components) {
      if (!map.has(comp.id)) {
        map.set(comp.id, { ...comp, paths: [...comp.paths] });
        continue;
      }
      const existing = map.get(comp.id);
      for (const p of comp.paths)  { if (!existing.paths.includes(p))  existing.paths.push(p); }
      for (const s of comp.scopes) { if (!existing.scopes.includes(s)) existing.scopes.push(s); }
    }

    return [...map.values()];
  }

  // ─────────────────────────────
  // ENTRY
  // ─────────────────────────────
  static async getInstalled(startDir = process.cwd()) {
    this.inventoryData = [];

    const visited = new Set();
    const raw     = [];

    function walk(dir) {
      let entries;
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }

      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        if (["node_modules", ".git", ".ubel", "vendor", ".bundle"].includes(entry.name)) continue;

        const full = path.join(dir, entry.name);

        if (RubyBundlerScanner._isBundlerRoot(full)) {
          const key = path.resolve(full);
          if (!visited.has(key)) {
            visited.add(key);
            raw.push(...RubyBundlerScanner._scanProject(full));
          }
          continue;
        }

        walk(full);
      }
    }

    if (RubyBundlerScanner._isBundlerRoot(startDir)) {
      const key = path.resolve(startDir);
      if (!visited.has(key)) {
        visited.add(key);
        raw.push(...RubyBundlerScanner._scanProject(startDir));
      }
    }

    walk(startDir);

    const merged = this.mergeInventoryByPurl(raw);
    this._assignScopes(merged);

    this.inventoryData = merged;
    return merged.map(c => c.id);
  }
}

export default RubyBundlerScanner;

File: python/ubel/ruby_runner.py
"""
ruby_scanner.py — Ruby / Bundler scanner (getInstalled only).

Mirrors the JS RubyBundlerScanner:
  - Detects Bundler roots (Gemfile + Gemfile.lock)
  - Reads Gemfile.lock for the full resolved gem graph
  - Reads Gemfile for group classifications (dev/test → dev, else → prod)
  - Assigns scopes via BFS from Gemfile groups
  - Merges duplicate PURLs
  - PURL: pkg:gem/<name>@<version>
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Any, Dict, List, Set


class RubyBundlerScanner:

    # ------------------------------------------------------------------ #
    # PURL                                                                 #
    # ------------------------------------------------------------------ #

    
    def _gem_purl(self,name: str, version: str) -> str:
        return f"pkg:gem/{name.lower()}@{version or ''}"

    # ------------------------------------------------------------------ #
    # Detect Bundler root                                                  #
    # ------------------------------------------------------------------ #

    
    def _is_bundler_root(self,directory: str) -> bool:
        return (
            os.path.exists(os.path.join(directory, "Gemfile")) and
            os.path.exists(os.path.join(directory, "Gemfile.lock"))
        )

    # ------------------------------------------------------------------ #
    # Parse Gemfile.lock                                                   #
    # ------------------------------------------------------------------ #

    
    def _parse_gemfile_lock(self,lock_path: str) -> Dict[str, Any]:
        """Returns lowercase_name → { name, version, dependencies: [lowercase names] }"""
        try:
            content = Path(lock_path).read_text(encoding="utf-8")
        except OSError:
            return {}

        index: Dict[str, Any] = {}
        in_specs   = False
        current_gem = None

        for raw_line in content.splitlines():
            line    = raw_line
            trimmed = line.strip()

            # Section header detection
            if re.match(r"^(GEM|PATH|GIT)$", trimmed):
                in_specs    = False
                current_gem = None
                continue

            if trimmed == "specs:":
                in_specs    = True
                current_gem = None
                continue

            # Exit on new uppercase section (PLATFORMS, DEPENDENCIES, etc.)
            if re.match(r"^[A-Z]", line) and trimmed != "specs:":
                in_specs    = False
                current_gem = None
                continue

            if not in_specs:
                continue

            indent  = len(line) - len(line.lstrip())
            if not trimmed:
                continue

            if indent == 4:
                # Top-level gem: "    name (version)"
                m = re.match(r"^([A-Za-z0-9_.\-]+)\s+\(([^)]+)\)", trimmed)
                if not m:
                    current_gem = None
                    continue
                name    = m.group(1)
                version = m.group(2).split(", ")[0]
                key     = name.lower()
                current_gem = {"name": name, "version": version, "dependencies": []}
                index[key]  = current_gem

            elif indent >= 6 and current_gem is not None:
                # Dependency of current gem
                m = re.match(r"^([A-Za-z0-9_.\-]+)", trimmed)
                if m:
                    current_gem["dependencies"].append(m.group(1).lower())

        return index

    # ------------------------------------------------------------------ #
    # Parse Gemfile for group classification                               #
    # ------------------------------------------------------------------ #

    
    def _parse_gemfile_groups(self,gemfile_path: str) -> tuple:
        """Returns (prod: Set[str], dev: Set[str])"""
        prod: Set[str] = set()
        dev:  Set[str] = set()

        try:
            content = Path(gemfile_path).read_text(encoding="utf-8")
        except OSError:
            return prod, dev

        DEV_GROUPS = {"development", "test", "staging"}
        current_groups: List[str] = []

        for raw_line in content.splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue

            # group :development, :test do
            group_block = re.match(r"^group\s+(.*?)\s+do$", line)
            if group_block:
                current_groups = [
                    g.strip().lstrip(":").lower()
                    for g in group_block.group(1).split(",")
                ]
                continue

            if line == "end":
                current_groups = []
                continue

            # gem 'name', ...
            gem_line = re.match(r'^gem\s+[\'"]([^\'"]+)[\'\"](.*)', line)
            if not gem_line:
                continue

            gem_name = gem_line.group(1).lower()
            rest     = gem_line.group(2) or ""

            # Inline group: group: :test  OR  groups: [:development, :test]
            inline_group = re.search(r"groups?:\s*(\[?[^,\]]+\]?)", rest)
            groups = list(current_groups)
            if inline_group:
                g_str  = re.sub(r"[\[\]]", "", inline_group.group(1))
                extras = [g.strip().lstrip(":").lower() for g in g_str.split(",")]
                groups = list(set(groups + extras))

            if any(g in DEV_GROUPS for g in groups):
                dev.add(gem_name)
            else:
                prod.add(gem_name)

        return prod, dev

    # ------------------------------------------------------------------ #
    # Scan one Bundler project                                             #
    # ------------------------------------------------------------------ #

    
    def _scan_project(self,project_root: str) -> List[Dict[str, Any]]:
        index = self._parse_gemfile_lock(
            os.path.join(project_root, "Gemfile.lock")
        )
        if not index:
            return []

        components: List[Dict[str, Any]] = []

        for key, entry in index.items():
            name    = entry["name"]
            version = entry["version"]
            cid     = self._gem_purl(name, version)

            resolved_deps: List[str] = []
            for dep in entry["dependencies"]:
                resolved = index.get(dep)
                if resolved:
                    resolved_deps.append(
                        self._gem_purl(resolved["name"], resolved["version"])
                    )
                else:
                    resolved_deps.append(self._gem_purl(dep, ""))

            components.append({
                "id":           cid,
                "name":         key,
                "version":      version,
                "type":         "library",
                "license":      "unknown",
                "ecosystem":    "ruby",
                "state":        "undetermined",
                "scopes":       [],
                "dependencies": resolved_deps,
                "paths":        [project_root],
                "project_root": project_root,
            })

        return components

    # ------------------------------------------------------------------ #
    # Assign scopes                                                        #
    # ------------------------------------------------------------------ #

    
    def _assign_scopes(self,inventory: List[Dict[str, Any]]) -> None:
        by_id: Dict[str, Dict] = {c["id"]: c for c in inventory}
        name_idx: Dict[str, List[Dict]] = {}

        for comp in inventory:
            if not isinstance(comp.get("scopes"), list):
                comp["scopes"] = []
            name_idx.setdefault(comp["name"], []).append(comp)

        project_groups: Dict[str, List[Dict]] = {}
        for comp in inventory:
            project_groups.setdefault(comp["project_root"], []).append(comp)

        for project_root, comps in project_groups.items():
            prod, dev = self._parse_gemfile_groups(
                os.path.join(project_root, "Gemfile")
            )

            def propagate(names: Set[str], scope: str) -> None:
                queue = [
                    c for n in names
                    for c in name_idx.get(n, [])
                    if c["project_root"] == project_root
                ]
                visited: Set[str] = set()
                while queue:
                    c = queue.pop(0)
                    if c["id"] in visited:
                        continue
                    visited.add(c["id"])
                    if scope not in c["scopes"]:
                        c["scopes"].append(scope)
                    for dep_id in c["dependencies"]:
                        d = by_id.get(dep_id)
                        if d and d.get("project_root") == project_root:
                            queue.append(d)

            propagate(prod, "prod")
            propagate(dev,  "dev")

            # Fallback
            for c in comps:
                if not c["scopes"]:
                    c["scopes"].append("prod")

    # ------------------------------------------------------------------ #
    # Merge by PURL                                                        #
    # ------------------------------------------------------------------ #

    
    def merge_inventory_by_purl(self,components: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        merged: Dict[str, Dict] = {}
        for comp in components:
            cid = comp["id"]
            if cid not in merged:
                clone = dict(comp)
                clone["paths"]  = list(comp.get("paths", []))
                clone["scopes"] = list(comp.get("scopes", []))
                merged[cid] = clone
                continue
            existing = merged[cid]
            for p in comp.get("paths", []):
                if p not in existing["paths"]:
                    existing["paths"].append(p)
            for s in comp.get("scopes", []):
                if s not in existing["scopes"]:
                    existing["scopes"].append(s)
        return list(merged.values())

    # ------------------------------------------------------------------ #
    # ENTRY                                                                #
    # ------------------------------------------------------------------ #

    def get_installed(self, start_dir: str = ".") -> List[str]:
        self.inventory_data = []
        start_dir = os.path.abspath(start_dir)

        visited: Set[str] = set()
        raw: List[Dict[str, Any]] = []

        skip = {"node_modules", ".git", ".ubel", "vendor", ".bundle"}

        def walk(directory: str) -> None:
            try:
                entries = os.scandir(directory)
            except OSError:
                return
            with entries as it:
                for entry in it:
                    if not entry.is_dir(follow_symlinks=False):
                        continue
                    if entry.name in skip:
                        continue
                    full = entry.path
                    if self._is_bundler_root(full):
                        key = os.path.realpath(full)
                        if key not in visited:
                            visited.add(key)
                            raw.extend(self._scan_project(full))
                        continue
                    walk(full)

        if self._is_bundler_root(start_dir):
            key = os.path.realpath(start_dir)
            if key not in visited:
                visited.add(key)
                raw.extend(self._scan_project(start_dir))

        walk(start_dir)

        merged = self.merge_inventory_by_purl(raw)
        self._assign_scopes(merged)

        self.inventory_data = merged
        return [c["id"] for c in merged]